### PR TITLE
add custom jade blind oracle app

### PR DESCRIPTION
**Pull request by valerio-vaccaro**
# App Submission
This is the Jade Blind Oracle available at https://github.com/Blockstream/blind_pin_server dockerized and with a configuration website added.

### App name
Blind Oracle

### 256x256 SVG icon
https://svgshare.com/i/ta0.svg

### Gallery images
https://i.imgur.com/iKlNNvD.jpeg
https://i.imgur.com/zur06iA.jpeg
https://i.imgur.com/CUDfUfF.jpeg

### I have tested my app on:
- [X] [Umbrel OS on a Raspberry Pi 4](https://github.com/getumbrel/umbrel-os)
- [X] [Custom Umbrel install on Linux x86 and arm](https://github.com/getumbrel/umbrel#-installation)
- [X] Available on [community app store](https://github.com/valerio-vaccaro/umbrel-test-repo)

### CODE REVIEW

File: pinserver/exports.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export APP_PINSERVER_PORT="8096"
```
Comment (smolgrrr) [inline pinserver/exports.sh:1]: Are you happy for us to change this? Will conflict with Jellyfin unfortunately https://github.com/getumbrel/umbrel-apps/blob/master/jellyfin/umbrel-app.yml

**Review by smolgrrr — COMMENTED** [review]
Thanks for this submission. Excited to see this one come to the app-store.

Just made some minor changes to umbrel-app.yml.
We will host the icon/gallery so no need to specify those, and no need for release notes initially.

Everything seems to be working great for me too from what I can see without a Jade.

File: pinserver/exports.sh (hunk 1 above)
Comment (valerio-vaccaro) [reply pinserver/exports.sh:1]: do you have an available port? i can update also images to have the same port exposed in the image

Comment (valerio-vaccaro) [reply pinserver/exports.sh:1]: Looks like port 8097 is free, i can move app on this port.
In next days if ok i will move on this port and update docker images.

Comment (smolgrrr) [reply pinserver/exports.sh:1]: perfect, thanks

**Comment by nmfretz** [comment]
Super stoked to see this come in @valerio-vaccaro! Fantastic work.

I hear you're pushing a couple improvements in the next couple days. Ping us as soon as you're ready.

**Comment by nmfretz** [comment]
Hi @valerio-vaccaro, how are things looking on your end?

**Comment by valerio-vaccaro** [comment]
Doing some more test on updated app, after I will update the code. in the weekend I plan to update this merge request and release version 0.1.0.

**Comment by valerio-vaccaro** [comment]
I finally updated images and fix other minimal issues. Code is ready for be merged

**Comment by nmfretz** [comment]
@valerio-vaccaro @bitcoinhelp, taking a look at this now.

**Review by nmfretz — COMMENTED** [review]
Nice work @valerio-vaccaro @bitcoinhelp! This is working really nicely.

I've reviewed and left some comments/suggestions below. Let me know if anything is unclear.

Very excited to get this to the app store 🙌

File: pinserver/umbrel-app.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        Blind oracle needs a Tor or a Tailscale connection, if you decide to use 
+        Tailscale please enable Tailscale's MagicDNS.
+developer: Blockstream
+website: https://blockstream.com
+dependencies: []
+repo: https://github.com/Blockstream/blind_pin_server
+support: https://t.me/blockstream_jade
+port: 3344
+gallery:
+  - 1.jpg
```
[the remark this answers is not in the document]
Comment (bitcoinhelp) [reply pinserver/umbrel-app.yml:26]: Sure thing, we'll put a couple more of these together!

File: pinserver/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      - ${APP_DATA_DIR}/data/server_private_key.key:/server_private_key.key
+      - ${APP_DATA_DIR}/data/pins:/pins
+
+  web: 
+    image: tulipan81/pinserver_web:v0.0.36@sha256:a5214dbc14538478744a6a0c6bd40f03ab8ac2990abc9038139c18901064205b
+    depends_on: 
+      - tor
+    restart: on-failure
+    ports:
+      - $APP_PINSERVER_WEB_PORT:$APP_PINSERVER_WEB_PORT
```
[the remark this answers is not in the document]
Comment (valerio-vaccaro) [reply pinserver/docker-compose.yml:29]: If i share directly the web interface i can send to a third party e.g. my friends and let they use my pinserver without login in umbrel, if i use the app_proxy page i will need to login 
There are no  dangerous commands on this page, just the qrcode in order to configure jade with the persona pinserver

[the remark this answers is not in the document]
Comment (nmfretz) [reply pinserver/docker-compose.yml:29]: Ah got it, that is a great use-case.

We can keep this port exposed, but one consideration for you is that Docker's internal dns can essentially handle this for you without you needing to expose a port on the host and needing to port forward from your router.

Here's an example using a Cloudflare Tunnel, but other methods would work too:
- Install the Cloudflare Tunnel app on Umbrel https://apps.umbrel.com/app/cloudflared
- Have a Cloudflare account set up with your own domain added, like `valerio.io`
- Create a tunnel directly to the blind oracle web container:

> public hostname that people can reach the app at:
> pinserver.valerio.io
>
> service to expose:
> http://pinserver_web_1:8095

Now when people visit  pinserver.valerio.io they will bypass our app proxy and connect directly to the `web` container. Docker handles resolving `pinserver_web_1` to the internal IP of the container.

[the remark this answers is not in the document]
Comment (nmfretz) [reply pinserver/docker-compose.yml:29]: Another option to weigh the pros and cons of is disabling the inbuilt authentication layer of the app proxy, which you can do with the following env var in the `app_proxy` service:

`PROXY_AUTH_ADD: "false"`

File: pinserver/hooks/pre-start [hunk 4]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
+
+# Create secret if needed
+"${UMBREL_ROOT}/scripts/app" compose "${APP_ID}" up --detach web
+
+# Delay booting pinserver until the Tor Hidden Service is ready
+HIDDEN_SERVICE_FILE="${TOR_DATA_DIR}/app-${APP_ID}/hostname"
```
[the remark this answers is not in the document]
Comment (valerio-vaccaro) [reply pinserver/hooks/pre-start:7]: web container will update secret if needed so i will need to turn on before the node with pinserver or pinserver will start with non-update secret after installation.

File: pinserver/torrc.template [hunk 5]
```diff
@@ -0,0 +1,4 @@
+# pinserver Hidden Service
+HiddenServiceDir /data/app-$APP_ID-node
+HiddenServicePort $APP_PINSERVER_PORT pinserver_node_1:8096
+HiddenServicePort 80 pinserver_web_1:$APP_PINSERVER_WEB_PORT
```
[the remark this answers is not in the document]
Comment (valerio-vaccaro) [reply pinserver/torrc.template:4]: Yes i want let an external user connect to the blind oracle configuration page (web) without login so a third party can use the same pinserver.

File: pinserver/docker-compose.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (valerio-vaccaro) [reply pinserver/docker-compose.yml:29]: can i share 2 apps with 2 different ports in the same app_proxy section?

**Comment by valerio-vaccaro** [comment]
As a recap:
- the app now has name "blockstream-blind-oracle"
- web interface is share via app_proxy without authentication (I want a third party be able to use my pinserver)
- pinserver (node) is shared by a custom tor (web and pinserver are two different docker applications)
- descriptions in umbrel-app.yaml are corrected as requested
- http/port in url for onion address are present because this is the format accepted by the jade and embedded in the qrcode 

I tested this new commit on umbrel 1.0 and previous and looks good in both. Let me know if something still needs some fixes!

**Comment by nmfretz** [comment]
wonderful @valerio-vaccaro, I will test this later today!

File: pinserver/docker-compose.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (nmfretz) [reply pinserver/docker-compose.yml]: unfortunately no. the app_proxy can only handle one port

**Comment by nmfretz** [comment]
Excellent, thanks @valerio-vaccaro. I have made the following changes to finalize this:

1. added reference to the new gallery images so they will render in the app store:  https://github.com/getumbrel/umbrel-apps/pull/950/commits/eeec326364f1c31b6652b1c83cb1162f2d4796f7

2. Modified the pre-start script https://github.com/getumbrel/umbrel-apps/pull/950/commits/4c222525f4d790fcfac46a4f2acab47642be1dc6 https://github.com/getumbrel/umbrel-apps/pull/950/commits/7e5807d2c836f792ba140da47a1190d988d75be5:
I made the pre-start script executable (`chmod +x`) so that it runs. It wasn't actually running before. I also modified the script it so that the web container gets the node container's hidden service env var on initial install. Without this a user would install the app and see http://notyetset.onion for the tor url until they restart the app.

This one is tricky to catch in testing because uninstalling an app doesn't remove the hidden service files at `~/umbrel/tor/data`. So you likely would not have run in to this issue regularly because the node container's hidden service file would exist on your umbrel after your initial test. It would then appear that everything is working even though the pre-start script is never running.

If you want to recreate this just so you can see that it is the case:
- uninstall the app
- remove the node containers hidden service file with `sudo rm -r ~/umbrel/tor/data/app-blockstream-blind-oracle-node`. You are now at a completely clean slate for the app.
- install the app (use the previous version, not the new one with my recent changes)
- go to the the app's UI and you will see http://notyetset.onion for the tor url. You can see that the web container doesn't have the necessary hidden service environment variable by running:
```
sudo docker exec <web-container-name> env
```

To get this to work on initial install you have to use the pre-start script to start the node container and the tor container so that a hidden service file can be created for the node container.


 It previously wasn't running  https://github.com/getumbrel/umbrel-apps/pull/950/commits/4c222525f4d790fcfac46a4f2acab47642be1dc6
skdgjhasdkljhfg 

Can you please check these over and make sure I haven't inadvertently changed some behaviour that you require. Once that's done this is good to go!

File: pinserver/hooks/pre-start (hunk 4 above)
[the remark this answers is not in the document]
Comment (nmfretz) [reply pinserver/hooks/pre-start]: Thanks for that explanation. I've made some changes to make the script executable and ensure that the web container gets the required env var while still updating the secret if needed. Explanation here: https://github.com/getumbrel/umbrel-apps/pull/950#issuecomment-2069456290

File: pinserver/umbrel-app.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (nmfretz) [reply pinserver/umbrel-app.yml]: Thanks for these!

**Comment by nmfretz** [comment]
Wonderful @valerio-vaccaro, thanks again for all your hard work on this! Blockstream Blind Oracle is now live on the app store 🎉.

The icon and gallery images live here: https://github.com/getumbrel/umbrel-apps-gallery/tree/master/blockstream-blind-oracle. We made a tiny tweak to the font.

**Comment by valerio-vaccaro** [comment]
Thanks a lot!

**Pull request merged**